File: android/src/main/java/com/reveila/android/VaultScanner.java
package com.reveila.android;

import android.content.Context;
import android.net.Uri;
import android.util.Log;
import androidx.documentfile.provider.DocumentFile;
import com.tom_roush.pdfbox.android.PDFBoxResourceLoader;
import com.tom_roush.pdfbox.io.MemoryUsageSetting;
import com.tom_roush.pdfbox.pdmodel.PDDocument;
import com.tom_roush.pdfbox.text.PDFTextStripper;
import org.w3c.dom.Document;
import org.w3c.dom.NodeList;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import com.reveila.android.ScanResult;
import com.reveila.android.data.VaultRepository;

public class VaultScanner {
    private static final String TAG = "VaultScanner";
    private static final long MAX_PDF_RAM_SIZE = 20 * 1024 * 1024L; // 20MB
    private final Context context;

    public VaultScanner(Context context) {
        this.context = context;
        PDFBoxResourceLoader.init(context);
    }

    /**
     * Headless scan logic. Synchronous execution (call from background thread).
     */
    public ScanResult performScan(
            Uri uri, 
            VaultRepository repository, // Replaces ViewModel
            String keywords,
            ScanProgressListener listener
    ) {
        List<String> userKeywords = new ArrayList<>();
        if (keywords != null && !keywords.isEmpty()) {
            for (String k : keywords.split(",")) {
                userKeywords.add(k.trim().toLowerCase());
            }
        }

        DocumentFile rootDir = DocumentFile.fromTreeUri(context, uri);
        if (rootDir == null) return new ScanResult(0, 0);

        DocumentFile[] allFiles = rootDir.listFiles();
        List<DocumentFile> filesToProcess = new ArrayList<>();
        
        // --- DELTA OPTIMIZATION ---
        Map<String, Long> indexedFiles = repository != null ? repository.getIndexedFiles() : java.util.Collections.emptyMap();

        for (DocumentFile file : allFiles) {
            String name = file.getName() != null ? file.getName().toLowerCase() : "";
            if (name.endsWith(".pdf") || name.endsWith(".md") || name.endsWith(".txt") || name.endsWith(".docx")) {
                long lastModified = file.lastModified();
                Long storedLastModified = indexedFiles.get(file.getUri().toString());
                
                if (storedLastModified == null || lastModified > storedLastModified) {
                    filesToProcess.add(file);
                }
            }
        }

        if (filesToProcess.isEmpty()) {
            Log.i(TAG, "Vault is up to date.");
            if (listener != null) listener.onProgress(1.0f);
            return new ScanResult(0, 0);
        }

        int entitiesFound = 0;
        ReasoningEngine engine = new ReasoningEngine(context, "gemma-3-1b-int8.gguf");

        for (int i = 0; i < filesToProcess.size(); i++) {
            DocumentFile file = filesToProcess.get(i);
            String fileName = file.getName() != null ? file.getName() : "Unknown";
            
            // 1. Extraction
            String extractedText = extractTextFromUri(file.getUri(), file.length());
            if (extractedText == null) extractedText = "Unreadable content.";

            // 2. Reasoning
            String analysis = engine.prompt("Analyze entities in: " + extractedText);
            entitiesFound += analysis.split(",").length;

            // 3. Persistence (Headless)
            if (repository != null) {
                repository.insertFact("Knowledge Vault", "CONCEPT", "CONTAINS", fileName, "DOCUMENT", analysis);
                repository.markFileAsIndexed(file.getUri().toString(), file.lastModified());

                // Keyword Priority Match
                for (String keyword : userKeywords) {
                    if (fileName.toLowerCase().contains(keyword) || extractedText.toLowerCase().contains(keyword)) {
                        repository.insertFact(fileName, "DOCUMENT", "PRIORITY_MATCH", keyword.toUpperCase(), "CONCEPT", "Detected focus keyword: " + keyword);
                        entitiesFound++;
                    }
                }
            }

            if (listener != null) {
                float progress = (float) (i + 1) / filesToProcess.size();
                listener.onProgress(progress);
            }
            
            try { Thread.sleep(500); } catch (InterruptedException ignored) {} // Pacing
        }

        engine.shutdown();
        return new ScanResult(filesToProcess.size(), entitiesFound);
    }

    private String extractTextFromUri(Uri uri, long size) {
        String path = uri.getPath();
        if (path == null) return null;
        String fileName = path.toLowerCase();
        
        try {
            if (fileName.endsWith(".pdf")) return extractPdfText(uri, size);
            if (fileName.endsWith(".docx")) return extractDocxText(uri);
            return extractPlainText(uri);
        } catch (Exception e) {
            Log.e(TAG, "Extraction failed for " + uri, e);
            return null;
        }
    }

    private String extractPlainText(Uri uri) throws Exception {
        StringBuilder sb = new StringBuilder();
        try (InputStream is = context.getContentResolver().openInputStream(uri);
             BufferedReader reader = new BufferedReader(new InputStreamReader(is))) {
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line).append("\n");
            }
        }
        return sb.toString();
    }

    private String extractPdfText(Uri uri, long size) throws Exception {
        MemoryUsageSetting memSetting = size > MAX_PDF_RAM_SIZE ? 
            MemoryUsageSetting.setupTempFileOnly() : MemoryUsageSetting.setupMainMemoryOnly();

        try (InputStream is = context.getContentResolver().openInputStream(uri);
             PDDocument document = PDDocument.load(is, memSetting)) {
            PDFTextStripper stripper = new PDFTextStripper();
            return stripper.getText(document);
        }
    }

    private String extractDocxText(Uri uri) {
        StringBuilder sb = new StringBuilder();
        try (InputStream is = context.getContentResolver().openInputStream(uri);
             ZipInputStream zip = new ZipInputStream(is)) {
            ZipEntry entry;
            while ((entry = zip.getNextEntry()) != null) {
                if ("word/document.xml".equals(entry.getName())) {
                    DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
                    DocumentBuilder builder = factory.newDocumentBuilder();
                    Document doc = builder.parse(zip);
                    NodeList textNodes = doc.getElementsByTagName("w:t");
                    for (int i = 0; i < textNodes.getLength(); i++) {
                        sb.append(textNodes.item(i).getTextContent()).append(" ");
                    }
                    break;
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Docx parse failed", e);
        }
        return sb.toString();
    }
}

File: android/src/main/kotlin/com/reveila/android/VaultScanner.kt
package com.reveila.android

import android.content.Context
import android.net.Uri
import android.util.Log
import androidx.documentfile.provider.DocumentFile
import com.reveila.android.ui.models.NodeType
import com.reveila.android.ui.viewmodels.SovereignMemoryViewModel
import com.tom_roush.pdfbox.android.PDFBoxResourceLoader
import com.tom_roush.pdfbox.pdmodel.PDDocument
import com.tom_roush.pdfbox.text.PDFTextStripper
import com.tom_roush.pdfbox.io.MemoryUsageSetting
import kotlinx.coroutines.delay
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.zip.ZipInputStream
import javax.xml.parsers.DocumentBuilderFactory

/**
 * Result of a vault scan operation.
 */
data class ScanResult(
    val newFilesCount: Int = 0,
    val entitiesDiscoveredCount: Int = 0
)

/**
 * Shared logic for scanning the Knowledge Vault.
 */
class VaultScanner(private val context: Context) {
    private val TAG = "VaultScanner"
    
    // Threshold for switching to disk-backed memory management (20MB)
    private val MAX_PDF_RAM_SIZE = 20 * 1024 * 1024L 

    init {
        // Initialize PDFBox for Android
        PDFBoxResourceLoader.init(context)
    }

    suspend fun performScan(
        uri: Uri, 
        viewModel: SovereignMemoryViewModel?, 
        keywords: String,
        onProgress: (Float) -> Unit = {}
    ): ScanResult {
        val userKeywords = keywords.split(",").map { it.trim().lowercase() }.filter { it.isNotEmpty() }
        
        val rootDir = DocumentFile.fromTreeUri(context, uri) ?: return ScanResult()
        val allFiles = rootDir.listFiles().filter { 
            val name = it.name?.lowercase() ?: ""
            name.endsWith(".pdf") || name.endsWith(".md") || name.endsWith(".txt") || name.endsWith(".docx")
        }
        
        // --- DELTA OPTIMIZATION ---
        val indexedFiles = viewModel?.getIndexedFiles() ?: emptyMap()
        val filesToProcess = allFiles.filter { file ->
            val uriString = file.uri.toString()
            val lastModified = file.lastModified()
            val storedLastModified = indexedFiles[uriString]
            storedLastModified == null || lastModified > storedLastModified
        }
        
        if (filesToProcess.isEmpty()) {
            Log.i(TAG, "No new or modified files to index. Vault is up to date.")
            viewModel?.setScanProgress(1.0f)
            onProgress(1.0f)
            return ScanResult()
        }

        var entitiesFound = 0
        val filesFound = filesToProcess.size

        // Initialize the Reasoning Engine (Mock for now as in Activity)
        val engine = ReasoningEngine(context, "gemma-3-1b-int8.gguf")
        
        filesToProcess.forEachIndexed { index, file ->
            val fileName = file.name ?: "Unknown Document"
            val fileUri = file.uri
            val fileUriString = fileUri.toString()
            val fileTimestamp = file.lastModified()
            val fileSize = file.length()
            val fileNameLower = fileName.lowercase()
            
            Log.i(TAG, "Indexing: $fileName ($fileSize bytes)")
            
            // 1. Extraction (Real & Memory Safe)
            val extractedText = extractTextFromUri(fileUri, fileSize) ?: "Empty or unreadable content."
            
            // 2. Reasoning
            val analysis = engine.prompt("Analyze entities in: $extractedText")
            // Simple heuristic to count "entities" in the analysis for the stats
            entitiesFound += analysis.split(",").size
            
            // 3. Update Model
            viewModel?.let {
                it.insertFact("Knowledge Vault", NodeType.CONCEPT, "CONTAINS", fileName, NodeType.DOCUMENT, analysis)
                it.markFileAsIndexed(fileUriString, fileTimestamp)
                
                val matchedKeyword = userKeywords.find { fileNameLower.contains(it) || extractedText.lowercase().contains(it) }
                if (matchedKeyword != null) {
                    it.insertFact(fileName, NodeType.DOCUMENT, "PRIORITY_MATCH", matchedKeyword.uppercase(), NodeType.CONCEPT, "Detected user-defined focus keyword: $matchedKeyword")
                    entitiesFound++
                }
            }

            val progress = (index + 1) / filesToProcess.size.toFloat()
            viewModel?.setScanProgress(progress)
            onProgress(progress)
            
            delay(500) // Pacing
        }
        
        engine.shutdown()
        viewModel?.setScanProgress(1.0f)
        onProgress(1.0f)
        
        return ScanResult(newFilesCount = filesFound, entitiesDiscoveredCount = entitiesFound)
    }

    private fun extractTextFromUri(uri: Uri, size: Long): String? {
        val fileName = uri.path?.lowercase() ?: ""
        return try {
            if (fileName.endsWith(".pdf")) {
                extractPdfText(uri, size)
            } else if (fileName.endsWith(".docx")) {
                extractDocxText(uri)
            } else {
                extractPlainText(uri)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to extract text from $uri", e)
            null
        }
    }

    private fun extractPlainText(uri: Uri): String {
        val stringBuilder = StringBuilder()
        context.contentResolver.openInputStream(uri)?.use { inputStream ->
            BufferedReader(InputStreamReader(inputStream)).use { reader ->
                var line: String?
                while (reader.readLine().also { line = it } != null) {
                    stringBuilder.append(line).append("\n")
                }
            }
        }
        return stringBuilder.toString()
    }

    private fun extractPdfText(uri: Uri, size: Long): String {
        val memSetting = if (size > MAX_PDF_RAM_SIZE) {
            Log.w(TAG, "Large PDF detected ($size bytes). Using disk-buffered extraction.")
            MemoryUsageSetting.setupTempFileOnly()
        } else {
            MemoryUsageSetting.setupMainMemoryOnly()
        }

        context.contentResolver.openInputStream(uri)?.use { inputStream ->
            PDDocument.load(inputStream, memSetting).use { document ->
                val stripper = PDFTextStripper()
                return stripper.getText(document)
            }
        }
        return ""
    }

    /**
     * Lightweight extraction of text from .docx files without heavy dependencies.
     * Parses the word/document.xml within the OOXML ZIP structure.
     */
    private fun extractDocxText(uri: Uri): String {
        val stringBuilder = StringBuilder()
        try {
            context.contentResolver.openInputStream(uri)?.use { inputStream ->
                ZipInputStream(inputStream).use { zip ->
                    var entry = zip.nextEntry
                    while (entry != null) {
                        if (entry.name == "word/document.xml") {
                            val factory = DocumentBuilderFactory.newInstance()
                            val builder = factory.newDocumentBuilder()
                            val doc = builder.parse(zip)
                            val textNodes = doc.getElementsByTagName("w:t")
                            for (i in 0 until textNodes.length) {
                                stringBuilder.append(textNodes.item(i).textContent).append(" ")
                            }
                            break
                        }
                        entry = zip.nextEntry
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse .docx: ${e.message}")
        }
        return stringBuilder.toString()
    }
}
